File: src/protocol.ts
import { randomUUID } from "node:crypto";

export const PROTOCOL_VERSION = 2;
export const MIN_PROTOCOL_VERSION = 1;

export type ErrorCode =
  | "BAD_REQUEST"
  | "CLIENT_UPGRADE_REQUIRED"
  | "CLIENT_OFFLINE"
  | "WORKSPACE_REQUIRED"
  | "WORKSPACE_OFFLINE"
  | "TOOL_TIMEOUT"
  | "TOOL_CANCELLED"
  | "TOOL_FAILED"
  | "POLICY_BLOCKED"
  | "APPROVAL_DENIED"
  | "SENSITIVE_PATH"
  | "PATH_ESCAPE"
  | "CONFLICT"
  | "NOT_FOUND"
  | "UNSUPPORTED"
  | "AUTH_FAILED"
  | "DUPLICATE_REQUEST";

export type ClientCapabilities = {
  filesystem: boolean;
  git: boolean;
  shell: boolean;
  pty: boolean;
  sandbox: "native" | "best-effort" | "policy-only" | "none";
  semanticProviders: string[];
  idempotency: boolean;
  cancellation: boolean;
  approvals: boolean;
  approvalMemory?: boolean;
  hostPolicyExecution?: boolean;
  mcpHub?: boolean;
};

export type RegisterMessage = {
  type: "register";
  protocolVersion: number;
  token?: string;
  credentialId?: string;
  credentialSecret?: string;
  deviceId: string;
  deviceName?: string;
  workspaceId: string;
  workspaceName: string;
  projectRoot?: string;
  capabilities: ClientCapabilities;
};

export type ToolCallMessage = {
  type: "tool_call";
  protocolVersion: number;
  requestId: string;
  sessionId?: string;
  workspaceKey: string;
  tool: string;
  args: unknown;
  idempotencyKey?: string;
  deadline?: number;
};

export type ToolCancelMessage = {
  type: "tool_cancel";
  protocolVersion: number;
  requestId: string;
  reason?: string;
};

export type ToolResultMessage = {
  type: "tool_result";
  protocolVersion: number;
  requestId: string;
  ok: boolean;
  result?: unknown;
  errorCode?: ErrorCode;
  errorMessage?: string;
  metadata?: Record<string, unknown>;
};

export type ServerMessage =
  | ToolCallMessage
  | ToolCancelMessage
  | { type: "registered"; protocolVersion: number; serverCapabilities: Record<string, unknown> }
  | { type: "ping"; ts: number };

export function requestId() {
  return randomUUID();
}

export function normalizeError(error: unknown): { errorCode: ErrorCode; errorMessage: string } {
  const message = error instanceof Error ? error.message : String(error);
  const lower = message.toLowerCase();
  if (lower.includes("sensitive-path")) return { errorCode: "SENSITIVE_PATH", errorMessage: message };
  if (lower.includes("escapes project_root") || lower.includes("path escape") || lower.includes("unsafe patch path")) return { errorCode: "PATH_ESCAPE", errorMessage: message };
  if (lower.includes("changed since read") || lower.includes("hash mismatch") || lower.includes("conflict")) return { errorCode: "CONFLICT", errorMessage: message };
  if (lower.includes("approval") && (lower.includes("denied") || lower.includes("rejected"))) return { errorCode: "APPROVAL_DENIED", errorMessage: message };
  if (lower.includes("blocked") || lower.includes("policy")) return { errorCode: "POLICY_BLOCKED", errorMessage: message };
  if (lower.includes("not found") || lower.includes("enoent") || lower.includes("unknown process")) return { errorCode: "NOT_FOUND", errorMessage: message };
  if (lower.includes("unsupported") || lower.includes("not available")) return { errorCode: "UNSUPPORTED", errorMessage: message };
  if (lower.includes("cancel")) return { errorCode: "TOOL_CANCELLED", errorMessage: message };
  return { errorCode: "TOOL_FAILED", errorMessage: message };
}

export function protocolCompatible(version: unknown) {
  return typeof version === "number" && version >= MIN_PROTOCOL_VERSION && version <= PROTOCOL_VERSION;
}

export function isSideEffectingTool(tool: string) {
  return new Set([
    "write_file",
    "edit_file",
    "apply_patch",
    "apply_edits",
    "format_changed_files",
    "run_command",
    "exec_start",
    "pty_start",
    "process_write",
    "process_kill",
    "exec_cancel",
    "git_stage",
    "git_unstage",
    "git_commit",
    "git_push",
    "approval_revoke",
    "approval_reset",
    "mcp_call",
  ]).has(tool);
}

File: internal/protocol/protocol.go
package protocol

import "time"

const (
	Version    = 3
	MinVersion = 1
)

type BrowserCapabilities struct {
	Available       bool `json:"available"`
	IsolatedProfile bool `json:"isolatedProfile,omitempty"`
	Screenshots     bool `json:"screenshots,omitempty"`
	Devtools        bool `json:"devtools,omitempty"`
	AttachExisting  bool `json:"attachExisting,omitempty"`
}

type ComputerCapabilities struct {
	Available         bool   `json:"available"`
	Backend           string `json:"backend,omitempty"`
	WindowList        bool   `json:"windowList,omitempty"`
	ScreenCapture     bool   `json:"screenCapture,omitempty"`
	UITree            bool   `json:"uiTree,omitempty"`
	Pointer           bool   `json:"pointer,omitempty"`
	Keyboard          bool   `json:"keyboard,omitempty"`
	Clipboard         bool   `json:"clipboard,omitempty"`
	BackgroundControl bool   `json:"backgroundControl,omitempty"`
	SecureDesktop     bool   `json:"secureDesktop,omitempty"`
}

type AutomationCapabilities struct {
	Browser  BrowserCapabilities  `json:"browser"`
	Computer ComputerCapabilities `json:"computer"`
}

type Capabilities struct {
	Filesystem           bool                   `json:"filesystem"`
	Git                  bool                   `json:"git"`
	Shell                bool                   `json:"shell"`
	PTY                  bool                   `json:"pty"`
	Sandbox              string                 `json:"sandbox"`
	SemanticProviders    []string               `json:"semanticProviders"`
	Idempotency          bool                   `json:"idempotency"`
	Cancellation         bool                   `json:"cancellation"`
	Approvals            bool                   `json:"approvals"`
	ApprovalMemory       bool                   `json:"approvalMemory,omitempty"`
	HostPolicyExecution  bool                   `json:"hostPolicyExecution,omitempty"`
	MCPHub               bool                   `json:"mcpHub,omitempty"`
	TerminalChatApproval bool                   `json:"terminalChatApproval,omitempty"`
	TerminalHistory      bool                   `json:"terminalHistory,omitempty"`
	Automation           AutomationCapabilities `json:"automation,omitempty"`
}

type RegisterMessage struct {
	Type             string       `json:"type"`
	ProtocolVersion  int          `json:"protocolVersion"`
	ClientVersion    string       `json:"clientVersion,omitempty"`
	Token            string       `json:"token,omitempty"`
	CredentialID     string       `json:"credentialId,omitempty"`
	CredentialSecret string       `json:"credentialSecret,omitempty"`
	DeviceID         string       `json:"deviceId"`
	DeviceName       string       `json:"deviceName,omitempty"`
	WorkspaceID      string       `json:"workspaceId"`
	WorkspaceName    string       `json:"workspaceName"`
	ProjectRoot      string       `json:"projectRoot,omitempty"`
	Capabilities     Capabilities `json:"capabilities"`
}

type ToolCall struct {
	Type            string         `json:"type"`
	ProtocolVersion int            `json:"protocolVersion"`
	RequestID       string         `json:"requestId"`
	SessionID       string         `json:"sessionId,omitempty"`
	WorkspaceKey    string         `json:"workspaceKey"`
	Tool            string         `json:"tool"`
	Args            map[string]any `json:"args,omitempty"`
	IdempotencyKey  string         `json:"idempotencyKey,omitempty"`
	Deadline        int64          `json:"deadline,omitempty"`
}

func (m ToolCall) DeadlineTime() (time.Time, bool) {
	if m.Deadline <= 0 {
		return time.Time{}, false
	}
	return time.UnixMilli(m.Deadline), true
}

type ToolCancel struct {
	Type            string `json:"type"`
	ProtocolVersion int    `json:"protocolVersion"`
	RequestID       string `json:"requestId"`
	Reason          string `json:"reason,omitempty"`
}

type ToolResult struct {
	Type            string `json:"type"`
	ProtocolVersion int    `json:"protocolVersion"`
	RequestID       string `json:"requestId"`
	OK              bool   `json:"ok"`
	Result          any    `json:"result,omitempty"`
	ErrorCode       string `json:"errorCode,omitempty"`
	ErrorMessage    string `json:"errorMessage,omitempty"`
	Metadata        any    `json:"metadata,omitempty"`
}

func Compatible(v int) bool { return v >= MinVersion && v <= Version }

func SideEffecting(tool string) bool {
	switch tool {
	case "write_file", "edit_file", "apply_patch", "apply_edits", "format_changed_files",
		"run_command", "exec_start", "pty_start", "process_write", "process_kill", "exec_cancel",
		"git_stage", "git_unstage", "git_commit", "git_push", "approval_revoke", "approval_reset", "mcp_call",
		"browser_open", "browser_click", "browser_fill", "browser_press", "browser_close",
		"computer_focus", "computer_click", "computer_type", "computer_key", "computer_scroll", "computer_drag":
		return true
	default:
		return false
	}
}
